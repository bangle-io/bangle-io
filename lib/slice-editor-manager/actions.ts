import type { BangleEditor } from '@bangle.dev/core';

import type { EditorIdType } from '@bangle.io/shared-types';

import { nsmEditorManagerSlice } from './slice';
import type { EditorSliceState } from './types';
import type { SelectionJson } from './utils';
import { isValidEditorId } from './utils';

export const noOp = nsmEditorManagerSlice.action(function noOp() {
  return nsmEditorManagerSlice.tx((state) => {
    return nsmEditorManagerSlice.update(state, (sliceState) => {
      return sliceState;
    });
  });
});

export const setEditingAllowed =
  nsmEditorManagerSlice.simpleAction('editingAllowed');

export const setEditor = nsmEditorManagerSlice.action(function setEditor({
  editorId,
  editor,
}: {
  editor: BangleEditor | undefined;
  editorId: EditorIdType;
}) {
  return nsmEditorManagerSlice.tx((state) => {
    return nsmEditorManagerSlice.update(state, (sliceState) => {
      if (!isValidEditorId(editorId)) {
        return sliceState;
      }

      const newEditors: EditorSliceState['mainEditors'] = [
        ...sliceState.mainEditors,
      ];
      newEditors[editorId] = editor;

      const editorOpenOrder = [editorId, ...sliceState.editorOpenOrder].filter(
        (index) => {
          return Boolean(newEditors[index]);
        },
      );

      return {
        mainEditors: newEditors,
        editorOpenOrder,
      };
    });
  });
});

export const onFocusUpdate = nsmEditorManagerSlice.simpleAction(
  'focusedEditorId',
  (editorId, state) => {
    if (typeof editorId === 'number' && !isValidEditorId(editorId)) {
      return nsmEditorManagerSlice.get(state);
    }

    return {
      focusedEditorId: editorId,
    };
  },
);

export const updateScrollPosition = nsmEditorManagerSlice.action(
  function updateScrollPosition({
    editorId,
    wsPath,
    scrollPosition,
  }: {
    editorId: EditorIdType;
    wsPath: string;
    scrollPosition: number;
  }) {
    return nsmEditorManagerSlice.tx((state) => {
      return nsmEditorManagerSlice.update(state, (sliceState) => {
        if (!isValidEditorId(editorId)) {
          return sliceState;
        }

        const newEditorConfig = sliceState.editorConfig.updateScrollPosition(
          scrollPosition,
          wsPath,
          editorId,
        );

        return {
          editorConfig: newEditorConfig,
        };
      });
    });
  },
);

export const incrementDisableEditingCounter = nsmEditorManagerSlice.action(
  function incrementDisableEditingCounter() {
    return nsmEditorManagerSlice.tx((state) => {
      return nsmEditorManagerSlice.update(state, (sliceState) => {
        return {
          disableEditingCounter: (sliceState.disableEditingCounter || 0) + 1,
        };
      });
    });
  },
);

export const updateSelection = nsmEditorManagerSlice.action(
  function updateSelection({
    editorId,
    wsPath,
    selectionJson,
  }: {
    editorId: EditorIdType;
    wsPath: string;
    selectionJson: SelectionJson;
  }) {
    return nsmEditorManagerSlice.tx((state) => {
      if (!isValidEditorId(editorId)) {
        return nsmEditorManagerSlice.get(state);
      }

      return nsmEditorManagerSlice.update(state, (sliceState) => {
        const newEditorConfig = sliceState.editorConfig.updateSelection(
          selectionJson,
          wsPath,
          editorId,
        );

        return {
          editorConfig: newEditorConfig,
        };
      });
    });
  },
);

/**
 * Enables or disables editing.
 *
 * @param param0.focusOrBlur - if true : focuses when enabling editing, blurs when disabling editing
 * @returns
 */
export const toggleEditing = nsmEditorManagerSlice.action(
  ({
    focusOrBlur = true,
    editingAllowed,
  }: { focusOrBlur?: boolean; editingAllowed?: boolean } = {}) => {
    return nsmEditorManagerSlice.tx((storeState) => {
      if (!focusOrBlur) {
        // NOOP
        return nsmEditorManagerSlice.get(storeState);
      }

      return nsmEditorManagerSlice.update(storeState, (sliceState) => {
        const newEditingAllowed = editingAllowed ?? !sliceState.editingAllowed;

        for (const editor of sliceState.mainEditors) {
          if (editor?.view.isDestroyed) {
            continue;
          }

          if (newEditingAllowed) {
            editor?.focusView();
          } else {
            editor?.view.dom.blur();
          }

          // send empty transaction so that editor view can update
          // the editing state.
          // queue it up so it is dispatched after store updated, so that
          // when editor queries the state, the state is correct.
          queueMicrotask(() => {
            editor?.view.dispatch(
              editor.view.state.tr.setMeta('__activitybar_empty__', true),
            );
          });
        }

        return {
          editingAllowed: newEditingAllowed,
        };
      });
    });
  },
);

export const updateQueryAction =
  nsmEditorManagerSlice.simpleAction('searchQuery');
