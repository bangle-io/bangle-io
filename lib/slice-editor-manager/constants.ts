import { MAX_OPEN_EDITORS } from '@bangle.io/constants';
import { createEmptyArray } from '@bangle.io/mini-js-utils';
import { checkWidescreen } from '@bangle.io/utils';

import { OpenedEditorsConfig } from './opened-editors-config';
import type { EditorSliceState } from './types';

// The time to wait before auto focusing any newly
// mounted editor
export const FOCUS_EDITOR_ON_LOAD_COOLDOWN = 1500;

export const initialEditorSliceState: EditorSliceState = {
  focusedEditorId: undefined,
  mainEditors: createEmptyArray(MAX_OPEN_EDITORS),
  editorConfig: OpenedEditorsConfig.fromJsonObj({
    selections: [],
    scrollPositions: [],
  }),
  primaryEditor: undefined,
  secondaryEditor: undefined,

  // We disable editing in mobile devices by default.
  // TODO: move this to a config stating the widescreen status
  editingAllowed: checkWidescreen(),

  editorOpenOrder: [],
  disableEditingCounter: undefined,
  searchQuery: undefined,
};
