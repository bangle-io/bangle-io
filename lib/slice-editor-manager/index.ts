export {
  onFocusUpdate,
  setEditor,
  toggleEditing,
  updateQueryAction,
  updateSelection,
} from './actions';
export { nsmEditorEffects } from './effects';
export {
  focusEditorIfNotFocused,
  forEachEditor,
  getEditor,
  getInitialSelection,
  nsmEditorManagerSlice,
  persistState,
  setEditorScrollPos,
} from './slice';
export type { EditorIdType } from './types';
