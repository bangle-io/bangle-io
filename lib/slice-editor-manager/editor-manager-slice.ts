import { isMobile } from '@bangle.io/config';
import { MAX_OPEN_EDITORS } from '@bangle.io/constants';
import { Slice } from '@bangle.io/create-store';
import { assertActionName, createEmptyArray } from '@bangle.io/utils';

import { editorManagerSliceKey } from './constants';
import {
  focusEditorEffect,
  initialSelectionEffect,
  trimWhiteSpaceEffect,
  watchEditorScrollEffect,
} from './effects';
import { OpenedEditorsConfig } from './opened-editors-config';
import type { EditorManagerAction, EditorSliceState } from './types';
import { calculateScrollPosition, calculateSelection } from './utils';

const LOG = false;
let log = LOG ? console.log.bind(console, 'editorManagerSlice') : () => {};

export const JSON_SCHEMA_VERSION = 'editor-slice/2';

export const initialEditorSliceState: EditorSliceState = {
  focusedEditorId: undefined,
  editors: createEmptyArray(MAX_OPEN_EDITORS),
  editorConfig: OpenedEditorsConfig.fromJsonObj({
    selections: [],
    scrollPositions: [],
  }),
  primaryEditor: undefined,
  secondaryEditor: undefined,
  // for now disabling editing for mobile by default
  // as it causes a lot of jumps due to keyboard
  editingAllowed: !isMobile,
};

const applyState = (
  action: EditorManagerAction,
  state: EditorSliceState,
): EditorSliceState => {
  switch (action.name) {
    case 'action::@bangle.io/slice-editor-manager:toggle-editing': {
      return {
        ...state,
        editingAllowed: !state.editingAllowed,
      };
    }
    case 'action::@bangle.io/slice-editor-manager:set-editor': {
      const { editorId, editor } = action.value;
      if (editorId >= MAX_OPEN_EDITORS) {
        throw new Error('editorId is out of range');
      }

      const newEditors: EditorSliceState['editors'] = [...state.editors];

      newEditors[editorId] = editor;

      return {
        ...state,
        editors: newEditors,
      };
    }

    case 'action::@bangle.io/slice-editor-manager:on-focus-update': {
      const editorId = action.value.editorId;
      if (typeof editorId === 'number' && editorId >= MAX_OPEN_EDITORS) {
        throw new Error('editorId is out of range');
      }

      return {
        ...state,
        focusedEditorId: editorId,
      };
    }

    case 'action::@bangle.io/slice-editor-manager:update-scroll-position': {
      const { editorId, wsPath, scrollPosition } = action.value;

      const newEditorConfig = state.editorConfig.updateScrollPosition(
        scrollPosition,
        wsPath,
        editorId,
      );

      return {
        ...state,
        editorConfig: newEditorConfig,
      };
    }

    case 'action::@bangle.io/slice-editor-manager:update-initial-selection-json': {
      const { editorId, wsPath, selectionJson } = action.value;

      const newEditorConfig = state.editorConfig.updateSelection(
        selectionJson,
        wsPath,
        editorId,
      );

      return {
        ...state,
        editorConfig: newEditorConfig,
      };
    }

    default:
      return state;
  }
};

export function editorManagerSlice(): Slice<
  EditorSliceState,
  EditorManagerAction
> {
  assertActionName('@bangle.io/slice-editor-manager', editorManagerSliceKey);

  return new Slice({
    key: editorManagerSliceKey,
    state: {
      init() {
        return initialEditorSliceState;
      },
      apply(action, state) {
        const newState = applyState(action, state);
        if (newState === state) {
          return state;
        }

        // derived state
        newState.primaryEditor = newState.editors[0];
        newState.secondaryEditor = newState.editors[1];

        return newState;
      },
      stateFromJSON(_, value: any) {
        if (!value || value.version !== JSON_SCHEMA_VERSION) {
          return initialEditorSliceState;
        }

        const data = value.data;

        const sliceState: EditorSliceState = Object.assign(
          {},
          initialEditorSliceState,
          {
            editorConfig: OpenedEditorsConfig.fromJsonObj(data.editorConfig),
            focusedEditorId: data.focusedEditorId,
          },
        );
        return sliceState;
      },

      stateToJSON(sliceState) {
        let newEditorConfig = sliceState.editorConfig;
        for (let i = 0; i < MAX_OPEN_EDITORS; i++) {
          const editor = sliceState.editors[i];
          if (editor) {
            const { editorId, selectionJson, wsPath } = calculateSelection(
              i,
              editor,
            );

            const scroll = calculateScrollPosition(i, editor)?.scrollPosition;
            newEditorConfig = newEditorConfig.updateSelection(
              selectionJson,
              wsPath,
              editorId,
            );
            newEditorConfig = newEditorConfig.updateScrollPosition(
              scroll,
              wsPath,
              editorId,
            );
          }
        }

        return {
          version: JSON_SCHEMA_VERSION,
          data: {
            ...initialEditorSliceState,
            editors: [],
            focusedEditorId: sliceState.focusedEditorId,
            editorConfig: newEditorConfig.toJsonObj(),
            primaryEditor: undefined,
            secondaryEditor: undefined,
          },
        };
      },
    },
    sideEffect: [
      initialSelectionEffect,
      focusEditorEffect,
      watchEditorScrollEffect,
      trimWhiteSpaceEffect,
    ],
  });
}
