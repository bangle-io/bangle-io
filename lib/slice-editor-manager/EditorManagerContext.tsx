import React, { useContext, useMemo } from 'react';

import {
  initialBangleStore,
  useSliceState,
} from '@bangle.io/bangle-store-context';
import { ApplicationStore } from '@bangle.io/create-store';

import { editorManagerSliceKey } from './constants';
import { initialEditorSliceState } from './editor-manager-slice';
import type { EditorSliceState } from './types';

const LOG = false;
let log = LOG ? console.log.bind(console, 'EditorManager') : () => {};

export type EditorManagerContextValue = EditorSliceState & {
  bangleStore: ApplicationStore;
};

// type EditorsType = [BangleEditor | undefined, BangleEditor | undefined];
const EditorManagerContext = React.createContext<EditorManagerContextValue>({
  ...initialEditorSliceState,
  bangleStore: initialBangleStore,
});

export function useEditorManagerContext() {
  return useContext(EditorManagerContext);
}

/**
 * Should be parent of all editors.
 */
export function EditorManager({ children }: { children: React.ReactNode }) {
  const { sliceState: editorManager, store } = useSliceState(
    editorManagerSliceKey,
  );

  const value = useMemo(() => {
    return {
      ...(editorManager || initialEditorSliceState),
      bangleStore: store,
    };
  }, [store, editorManager]);

  return (
    <EditorManagerContext.Provider value={value}>
      {children}
    </EditorManagerContext.Provider>
  );
}
