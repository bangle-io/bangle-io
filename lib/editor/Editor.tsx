import React, {
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';

import type { BangleEditor as CoreBangleEditor } from '@bangle.dev/core';
import type { Node, Selection } from '@bangle.dev/pm';
import type { RenderNodeViewsFunction } from '@bangle.dev/react';
import {
  BangleEditor as ReactBangleEditor,
  useEditorState,
} from '@bangle.dev/react';
import { valuePlugin } from '@bangle.dev/utils';

import {
  useNsmPlainStore,
  useNsmStore,
  useSerialOperationContext,
} from '@bangle.io/api';
import { EditorDisplayType } from '@bangle.io/constants';
import { vars } from '@bangle.io/css-vars';
import { EditorPluginMetadataKey } from '@bangle.io/editor-common';
import { nsmSliceWorkspace } from '@bangle.io/nsm-slice-workspace';
import type {
  DispatchSerialOperationType,
  EditorPluginMetadata,
  EternalVars,
  NsmStore,
  WsPath,
} from '@bangle.io/shared-types';
import type { EditorIdType } from '@bangle.io/slice-editor-manager';
import {
  getEditor,
  getInitialSelection,
  nsmEditorManagerSlice,
  setEditor,
  setEditorScrollPos,
  updateSelection,
} from '@bangle.io/slice-editor-manager';
import { calculateSelection } from '@bangle.io/slice-editor-manager/utils';
import { cx } from '@bangle.io/utils';
import { fs } from '@bangle.io/workspace-info';

import { watchPluginHost } from './watch-plugin-host';

const LOG = false;
let log = LOG ? console.log.bind(console, 'play/Editor') : () => {};

export interface EditorProps {
  className?: string;
  editorDisplayType?: EditorDisplayType;
  editorId: EditorIdType;
  eternalVars: EternalVars;
  wsPath: WsPath;
}

export function Editor(props: EditorProps) {
  // using key=wsPath to force unmount and mount the editor
  // with fresh values
  return <EditorInner key={props.wsPath} {...props} />;
}

function EditorInner({
  className,
  editorDisplayType = EditorDisplayType.Page,
  editorId,
  eternalVars,
  wsPath,
}: EditorProps) {
  const { dispatchSerialOperation } = useSerialOperationContext();
  const editorStore = useNsmStore([nsmEditorManagerSlice, nsmSliceWorkspace]);
  const nsmStore = useNsmPlainStore();
  // Even though the collab extension will reset the content to its convenience
  // preloading the content will give us the benefit of static height, which comes
  // in handy when loading editor with a given scroll position.
  const [initialValue, setInitialDoc] = useState<Node | undefined>();

  useEffect(() => {
    let destroyed = false;

    fs.getNote(wsPath, eternalVars.extensionRegistry).then((doc) => {
      if (!destroyed) {
        setInitialDoc(doc);
      }
    });

    return () => {
      destroyed = true;
    };
  }, [wsPath, eternalVars.extensionRegistry]);

  const editorRef = useRef<ReturnType<typeof Proxy.revocable> | null>(null);

  const _onEditorReady = useCallback(
    (editor) => {
      const proxiedEditor = Proxy.revocable(editor, {});
      editorRef.current = proxiedEditor;

      setEditorScrollPos(editorStore.state, wsPath, editorId);
      editorStore.dispatch(
        setEditor({
          editorId,
          editor: proxiedEditor.proxy,
        }),
      );
    },
    [editorStore, wsPath, editorId],
  );

  useEffect(() => {
    return () => {
      const editorProxy = editorRef.current;

      if (editorProxy) {
        editorRef.current = null;

        const currentEditor = getEditor(editorStore.state, editorId);

        // make sure we are unsetting the correct editor
        if (currentEditor === editorProxy.proxy) {
          editorStore.dispatch(
            setEditor({
              editor: undefined,
              editorId,
            }),
          );
          // update the last selection before unmounting
          const value = calculateSelection(editorId, currentEditor);
          editorStore.dispatch(updateSelection(value));
        }

        // Avoiding MEMORY LEAK
        // Editor object is a pretty massive object and writing idiomatic react
        // makes you use caching interfaces like useMemo, React.Memo etc, which
        // cache their dependencies. Caching majority of the things is fine, but
        // inadvertently caching an object like Editor causes major memory leaks.
        // There are various ways to avoid this problem like avoiding the usage of
        // editor in React components, or just being extra careful.
        // For the scope of this project, being careful is prone to errors, so a quick
        // and dirty approach is to put a revocable proxy in between rest of application
        // and the Editor instance.This works because we know for sure if an Editor has
        // been destroyed, we will never be reusing it, so we severe the reference by calling
        // `.revoke()` and let GC collect the Editor once it is destroyed. The timeout exists
        // just to give other places some time before the proxy is revoked.
        setTimeout(() => {
          editorProxy.revoke();
        }, 100);
      }
    };
  }, [editorId, editorStore]);

  const initialSelection =
    editorId != null && initialValue
      ? getInitialSelection(editorStore.state, editorId, wsPath, initialValue)
      : undefined;

  return initialValue ? (
    <EditorInner2
      nsmStore={nsmStore}
      initialSelection={initialSelection}
      dispatchSerialOperation={dispatchSerialOperation}
      className={className}
      editorId={editorId}
      eternalVars={eternalVars}
      initialValue={initialValue}
      wsPath={wsPath}
      onEditorReady={_onEditorReady}
      editorDisplayType={editorDisplayType}
    />
  ) : null;
}

function EditorInner2({
  nsmStore,
  className,
  dispatchSerialOperation,
  editorDisplayType,
  editorId,
  eternalVars,
  initialValue,
  onEditorReady,
  initialSelection,
  wsPath,
}: {
  eternalVars: EternalVars;
  nsmStore: NsmStore;
  className?: string;
  dispatchSerialOperation: DispatchSerialOperationType;
  editorDisplayType: EditorDisplayType;
  editorId?: EditorIdType;
  initialValue: any;
  initialSelection: Selection | undefined;
  onEditorReady?: (editor: CoreBangleEditor) => void;
  wsPath: WsPath;
}) {
  const editorState = useGetEditorState({
    dispatchSerialOperation,
    editorDisplayType,
    editorId,
    eternalVars,
    initialSelection,
    initialValue,
    wsPath,
    nsmStore,
  });

  const renderNodeViews: RenderNodeViewsFunction = useCallback(
    (nodeViewRenderArg) => {
      return eternalVars.extensionRegistry.renderReactNodeViews({
        nodeViewRenderArg,
      });
    },
    [eternalVars],
  );

  let displayClass = 'B-editor_display-page';
  switch (editorDisplayType) {
    case EditorDisplayType.Page: {
      displayClass = 'B-editor_display-page';
      break;
    }
    case EditorDisplayType.Floating: {
      displayClass = 'B-editor_display-popup';
      break;
    }
    default: {
      // hack to catch switch slipping
      let val: never = editorDisplayType;
      throw new Error('Unknown error type ' + val);
    }
  }

  return (
    <ReactBangleEditor
      className={cx(className, displayClass)}
      focusOnInit={false}
      onReady={onEditorReady}
      renderNodeViews={renderNodeViews}
      state={editorState}
    >
      {eternalVars.extensionRegistry.renderExtensionEditorComponents()}
    </ReactBangleEditor>
  );
}

export function useGetEditorState({
  dispatchSerialOperation,
  editorDisplayType,
  editorId,
  initialSelection,
  initialValue,
  wsPath,
  nsmStore,
  eternalVars,
}: {
  nsmStore: NsmStore;
  dispatchSerialOperation: DispatchSerialOperationType;
  editorDisplayType: EditorDisplayType;
  editorId?: EditorIdType;
  eternalVars: EternalVars;
  initialSelection: Selection | undefined;
  initialValue: any;
  wsPath: WsPath;
}) {
  // TODO decouple pluginMetadata, this should be provided as a prop
  const pluginMetadata: EditorPluginMetadata = useMemo(
    () => ({
      wsPath,
      editorId,
      editorDisplayType,
      dispatchSerialOperation,
      nsmStore,
      createdAt: Date.now(),
      collabMessageBus: eternalVars.editorCollabMessageBus,
    }),

    [
      editorId,
      nsmStore,
      wsPath,
      dispatchSerialOperation,
      editorDisplayType,
      eternalVars,
    ],
  );

  const plugins = useCallback(() => {
    return [
      // needs to be at top so that other plugins get depend on this
      valuePlugin(EditorPluginMetadataKey, pluginMetadata),
      ...eternalVars.extensionRegistry.getPlugins(),

      // Needs to be at bottom so that it can dispatch
      // operations for any plugin state updates before it
      watchPluginHost(
        pluginMetadata,
        eternalVars.extensionRegistry.getEditorWatchPluginStates(),
      ),
    ];
  }, [eternalVars, pluginMetadata]);

  const editorState = useEditorState({
    plugins,
    pluginMetadata,
    specRegistry: eternalVars.extensionRegistry.specRegistry,
    initialValue,
    pmStateOpts: {
      selection: initialSelection,
    },
    editorProps: {},
    dropCursorOpts: {
      color: vars.color.promote.solidStrong,
      width: 2,
    },
  });

  return editorState;
}
