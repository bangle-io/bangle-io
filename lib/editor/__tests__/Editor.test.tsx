/**
 * @jest-environment @bangle.io/jsdom-env
 */
import { act, render, waitFor } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';

import { Selection } from '@bangle.dev/pm';

import {
  EditorDisplayType,
  PRIMARY_EDITOR_INDEX,
  SECONDARY_EDITOR_INDEX,
} from '@bangle.io/constants';
import type { EditorIdType } from '@bangle.io/slice-editor-manager';
import { getEditor, getEditorView } from '@bangle.io/slice-editor-manager';
import {
  createBasicTestStore,
  createExtensionRegistry,
  createPMNode,
  setupMockWorkspaceWithNotes,
  TestStoreProvider,
} from '@bangle.io/test-utils';

import { Editor, useGetEditorState } from '../Editor';

const extensionRegistry = createExtensionRegistry([], {
  editorCore: true,
});

let result: ReturnType<typeof render>;

const ONE_DOC_CONTENT = `Hello world! I am a test`;
const TWO_DOC_CONTENT = `bubble gum`;
const wsName = 'test-ws';

beforeEach(() => {
  jest.useRealTimers();
});

const setup = async () => {
  const noteWsPaths: Array<[string, string]> = [
    [`${wsName}:one.md`, ONE_DOC_CONTENT],
    [`${wsName}:two.md`, TWO_DOC_CONTENT],
  ];

  const { store } = createBasicTestStore({
    useEditorManagerSlice: true,
    useUISlice: true,
  });
  await setupMockWorkspaceWithNotes(store, wsName, noteWsPaths);

  return { store };
};

test('basic renders', async () => {
  const { store: bangleStore } = await setup();

  await act(async () => {
    result = render(
      <TestStoreProvider bangleStore={bangleStore} bangleStoreChanged={0}>
        <Editor
          editorId={SECONDARY_EDITOR_INDEX}
          wsPath="test-ws:one.md"
          className="test-class"
          extensionRegistry={extensionRegistry}
        />
      </TestStoreProvider>,
    );
  });

  await waitFor(() => {
    expect(result!.container.innerHTML).toContain(ONE_DOC_CONTENT);
  });

  await waitFor(() => {
    expect(result.container.innerHTML).toContain('class="test-class');
  });

  expect(result!.container.innerHTML).toContain(ONE_DOC_CONTENT);
  expect(result!.container).toMatchSnapshot();
});

test('persists editor selection', async () => {
  const setSelectionAtEnd = (editorId: EditorIdType) => {
    const view = getEditorView(editorId)(bangleStore.state)!;
    view.dispatch(view.state.tr.setSelection(Selection.atEnd(view.state.doc)));
  };

  const getEditorSelection = (editorId: EditorIdType) => {
    const view = getEditorView(editorId)(bangleStore.state)!;

    return view.state.selection.toJSON();
  };

  const { store: bangleStore } = await setup();

  await act(async () => {
    result = render(
      <TestStoreProvider bangleStore={bangleStore} bangleStoreChanged={0}>
        <Editor
          editorId={SECONDARY_EDITOR_INDEX}
          wsPath="test-ws:one.md"
          className="test-class1"
          extensionRegistry={extensionRegistry}
        />
      </TestStoreProvider>,
    );
  });

  await waitFor(() => {
    expect(result!.container.innerHTML).toContain(ONE_DOC_CONTENT);
  });

  expect(getEditorSelection(SECONDARY_EDITOR_INDEX)).toEqual({
    anchor: 1,
    head: 1,
    type: 'text',
  });

  // change the selection so we can test if selection is persisted
  // if we open the document again.
  setSelectionAtEnd(SECONDARY_EDITOR_INDEX);

  const expectedSelection = {
    anchor: 25,
    head: 25,
    type: 'text',
  };
  expect(getEditorSelection(SECONDARY_EDITOR_INDEX)).toEqual(expectedSelection);

  // load another document
  act(() => {
    result.rerender(
      <TestStoreProvider bangleStore={bangleStore} bangleStoreChanged={0}>
        <Editor
          editorId={SECONDARY_EDITOR_INDEX}
          wsPath="test-ws:two.md"
          className="test-class2"
          extensionRegistry={extensionRegistry}
        />
      </TestStoreProvider>,
    );
  });

  await waitFor(() => {
    expect(result!.container.innerHTML).toContain(TWO_DOC_CONTENT);
  });
  // The new editor should not use the selection from one.md
  expect(getEditorSelection(SECONDARY_EDITOR_INDEX)).toEqual({
    anchor: 1,
    head: 1,
    type: 'text',
  });

  // now load the one.md document again, this time the selection
  // should be hydrated with `expectedSelection`
  act(() => {
    result = render(
      <TestStoreProvider bangleStore={bangleStore} bangleStoreChanged={0}>
        <Editor
          editorId={SECONDARY_EDITOR_INDEX}
          wsPath="test-ws:one.md"
          className="test-class1"
          extensionRegistry={extensionRegistry}
        />
      </TestStoreProvider>,
    );
  });

  await waitFor(() => {
    expect(result!.container.innerHTML).toContain(ONE_DOC_CONTENT);
  });

  expect(getEditorSelection(SECONDARY_EDITOR_INDEX)).toEqual(expectedSelection);
});

test('mounting and unmounting set state correctly in editor slice', async () => {
  const { store: bangleStore } = await setup();

  await act(async () => {
    result = render(
      <TestStoreProvider bangleStore={bangleStore} bangleStoreChanged={0}>
        <Editor
          editorId={SECONDARY_EDITOR_INDEX}
          wsPath="test-ws:one.md"
          className="test-class"
          extensionRegistry={extensionRegistry}
        />
      </TestStoreProvider>,
    );
  });

  await waitFor(() => {
    expect(result!.container.innerHTML).toContain(ONE_DOC_CONTENT);
  });
  await waitFor(() => {
    expect(result.container.innerHTML).toContain('class="test-class');
  });

  expect(getEditor(PRIMARY_EDITOR_INDEX)(bangleStore.state)).toBeUndefined();
  expect(getEditor(SECONDARY_EDITOR_INDEX)(bangleStore.state)).toBeDefined();

  result.unmount();

  expect(getEditor(PRIMARY_EDITOR_INDEX)(bangleStore.state)).toBeUndefined();
  expect(getEditor(SECONDARY_EDITOR_INDEX)(bangleStore.state)).toBeUndefined();
});

test('revokes editor proxy', async () => {
  const { store: bangleStore } = await setup();

  let revokeSpy = jest.fn();
  let spy = jest
    .spyOn(global.Proxy, 'revocable')
    .mockImplementation((r) => ({ proxy: r, revoke: revokeSpy }));

  await act(async () => {
    result = render(
      <TestStoreProvider bangleStore={bangleStore} bangleStoreChanged={0}>
        <Editor
          editorId={SECONDARY_EDITOR_INDEX}
          wsPath="test-ws:one.md"
          className="test-class"
          extensionRegistry={extensionRegistry}
        />
      </TestStoreProvider>,
    );
  });

  await waitFor(() => {
    expect(result!.container.innerHTML).toContain(ONE_DOC_CONTENT);
  });

  await waitFor(() => {
    expect(result.container.innerHTML).toContain('class="test-class');
  });

  jest.useFakeTimers();

  act(() => {
    result.unmount();
  });

  jest.runAllTimers();

  expect(spy).toBeCalledTimes(1);

  expect(revokeSpy).toBeCalledTimes(1);
});

test('changing of wsPath works', async () => {
  const { store: bangleStore } = await setup();

  await act(async () => {
    result = render(
      <TestStoreProvider bangleStore={bangleStore} bangleStoreChanged={0}>
        <Editor
          editorId={SECONDARY_EDITOR_INDEX}
          className="test-class"
          wsPath="test-ws:one.md"
          extensionRegistry={extensionRegistry}
        />
      </TestStoreProvider>,
    );
  });

  await waitFor(() => {
    expect(result!.container.innerHTML).toContain(ONE_DOC_CONTENT);
  });

  await waitFor(() => {
    expect(result.container.innerHTML).toContain('class="test-class');
  });

  expect(result!.container.innerHTML).toContain(ONE_DOC_CONTENT);

  await act(async () => {
    result.rerender(
      <TestStoreProvider bangleStore={bangleStore} bangleStoreChanged={0}>
        <Editor
          editorId={SECONDARY_EDITOR_INDEX}
          wsPath="test-ws:two.md"
          className="test-class"
          extensionRegistry={extensionRegistry}
        />
      </TestStoreProvider>,
    );
  });

  await waitFor(() => {
    expect(result.container.innerHTML).toContain('class="test-class');
  });

  expect(result!.container.innerHTML).toContain(TWO_DOC_CONTENT);
});

describe('useGetEditorState', () => {
  test('generates correct state', async () => {
    const { store: bangleStore } = await setup();

    const { result } = renderHook(() =>
      useGetEditorState({
        editorId: PRIMARY_EDITOR_INDEX,
        extensionRegistry,
        initialValue: '',
        wsPath: 'something:one.md',
        editorDisplayType: EditorDisplayType.Page,
        dispatchSerialOperation: jest.fn(),
        bangleStore: bangleStore,
        initialSelection: undefined,
      }),
    );

    expect(result.current.pmState.toJSON()).toMatchInlineSnapshot(`
      {
        "doc": {
          "content": [
            {
              "type": "paragraph",
            },
          ],
          "type": "doc",
        },
        "selection": {
          "anchor": 1,
          "head": 1,
          "type": "text",
        },
      }
    `);
    expect(result.current.specRegistry).toBeTruthy();
  });

  test('when initial selection is provided', async () => {
    const pmNode = createPMNode([], `# Hello World`.trim());
    const { store: bangleStore } = await setup();

    const { result } = renderHook(() =>
      useGetEditorState({
        editorId: PRIMARY_EDITOR_INDEX,
        extensionRegistry,
        initialValue: pmNode,
        wsPath: 'something:one.md',
        editorDisplayType: EditorDisplayType.Page,
        dispatchSerialOperation: jest.fn(),
        bangleStore: bangleStore,
        initialSelection: Selection.fromJSON(pmNode, {
          anchor: 5,
          head: 5,
          type: 'text',
        }),
      }),
    );

    expect(result.current.pmState.toJSON().selection).toEqual({
      anchor: 5,
      head: 5,
      type: 'text',
    });

    expect(result.current.pmState.toJSON()).toMatchInlineSnapshot(`
      {
        "doc": {
          "content": [
            {
              "attrs": {
                "collapseContent": null,
                "level": 1,
              },
              "content": [
                {
                  "text": "Hello World",
                  "type": "text",
                },
              ],
              "type": "heading",
            },
          ],
          "type": "doc",
        },
        "selection": {
          "anchor": 5,
          "head": 5,
          "type": "text",
        },
      }
    `);
    expect(result.current.specRegistry).toBeTruthy();
  });
});
