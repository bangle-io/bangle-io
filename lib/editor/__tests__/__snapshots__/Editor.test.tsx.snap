// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic renders 1`] = `
<div>
  <div
    class="test-class B-editor_display-page "
  >
    <div
      class="ProseMirror bangle-editor"
      contenteditable="true"
      translate="no"
    >
      <p>
        Hello world! I am a test
      </p>
    </div>
  </div>
</div>
`;
